### ISSUE TITLE
Boxes with negative scores in NMS input?

### ISSUE THREAD

**Issue by masahi** [pytorch/vision#3198]
Hi, I found that the use of NMS in `RegionProposalNetwork` can take on boxes with negative scores as inputs. I found this when running MaskRCNN in v0.8 release.

https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/rpn.py#L261

In other use of NMS in `ROIHeads`, scores are thresholded to keep only boxes with positive scores:
https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/roi_heads.py#L703

I'm wondering if that lack of score thresholding in RPN is intentional or not... In TVM, we expects NMS input with negative scores to be invalid. Since NMS in PyTorch doesn't have a score threshold parameter, we didn't realize that there could be boxes with negative scores. 

I proposed to fix TVM's NMS conversion in https://github.com/apache/tvm/pull/7137, but since it would have a big performance implication and I heard that negative boxes don't matter in the final output anyway, I'm now inclined not to fix this in TVM side.

cc @fmassa  @t-vi

**Comment by datumbox** [comment]
I believe that the original intention of the following snippet was to speed up the execution of the NMS op rather than filtering explicitly any non positive value:
https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/roi_heads.py#L702-L704

Could you provide a snippet that reproduces negative scores so that we can investigate?

**Comment by masahi** [comment]
Sure this is the script https://gist.github.com/masahi/ea002c85e7d665d40eeb5c6422490e63

I dumped the inputs to `batched_nms` and I can see many negative boxes. My dumps are available at https://github.com/masahi/torchscript-to-tvm/tree/master/nms-perf-issue

I think thresholding scores at 0 make sense because the scores in `ROIHeads` are outputs from softmax that are supposed to be probability-like values. But in the other use of `batched_nms`, the one I'm getting many negative boxes from, the scores are "objectness". Objectness seems to be outputs from convolution layers whose values can be anything. Actually, when I run my script,  there are about 150000 boxes generated from objectness network, and only about 3000 of them have positive scores.

**Comment by datumbox** [comment]
Thanks a lot for providing additional info.

NMS does not require the scores to be positive. The kernel uses the scores to rank the boxes from highest to lowest. Thus they can be virtually any well-behaved score and the method will still work:
https://github.com/pytorch/vision/blob/7d831a2f9b3ebab9eb8e5c899cf70b103ad6908a/torchvision/csrc/ops/cpu/nms_kernel.cpp#L30

As you said, in the case of `ROIHeads` the scores are positive because they are probabilities:
https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/roi_heads.py#L677

Thus in this case there is a natural range and a lower threshold value that can be assumed. I think this is the reason why applying the heuristic thresholding is easier. 

On the other hand, the objectness score are logits so it takes values in R (a negative logit is when p <0.5). You can see that this is the case here:
https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/rpn.py#L301-L303

So in principle you could introduce such a threshold in the RPN method too, but you would have to turn logits into probabilities to ensure their scale is more intuitive for the user and they can tune easier the threshold. To reiterate, we will be doing this not because NMS does not handle negative scores, but because we might want to apply the same heuristic as in ROIHeads and clip boxes before NMS. Perhaps if we had a few benchmarks that indicate massive speed up gains with minimum loss in accuracy, you could build a strong case.

I'll like the opinion of @fmassa who is way more into Object Detection than me to see if that's something we want to bring in TorchVision. He is currently on holidays so, he is likely to respond once he is back.

**Comment by masahi** [comment]
ok I understand that negative boxes are considered valid inputs to NMS. But then there are some deep questions I have like the ones below. I don't expect all of them to be answered, since some of them might not have a clear answer.

* What do boxes with negative objectness mean?
* What makes one box positive while others negative?
* Can a totally reasonable box have a negative objectness? Some people in the TVM community are claiming that negative boxes don't matter in the final detection results in practice. 

If being negative doesn't matter much in terms of objectness, would it be better to rescale them in the range [0, 1)? If yes, I think it is better to introduce a user-chosen threshold parameter to cut low-objectness boxes. There is already `box_score_thresh` parameter for NMS use in `roi_head.py`, so I think having a similar tunable parameter for NMS with objectness in `rpn.py` would be reasonable.

**Comment by zhiqwang** [comment]
I'm not sure the negative box will cause this problem as described in #1705 ?

**Comment by masahi** [comment]
@zhiqwang No. If negative boxes are considered valid, we cannot reduce the number of boxes.

**Comment by datumbox** [comment]
@masahi I'm going to give a lengthier answer before jumping to your questions because it will give you a better understanding on how the implementation works:

In both RPN and RoIHeads we have parts of the network that do classification and estimate class [logits](https://en.wikipedia.org/wiki/Logit) instead of class probabilities (one for whether there is anything inside the anchor and the other about the category label). You can convert logits to probabilities and vice-versa easily but estimating logits is cheaper because you don't have to exponate, normalize etc. 

You can confirm that we work with class logits by looking at the loss functions used in the fasterrcnn components ([binary_cross_entropy_with_logits](https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/rpn.py#L301-L303) and [fastrcnn_loss](https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/roi_heads.py#L35)). In both the cases of RPN and RoiHeads those "scores" are fitted against some labels (not the same!). So in both cases, we estimate classification logits.

So why one technique ends up with negative scores? Well on `RoIHeads`, we do the following just before calling NMS and we turn the logits into probabilities: 
https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/models/detection/roi_heads.py#L677

On the other hand, in the case of RPN we use directly the logits (which on the code they are called objectness) without converting them to probabilities (you can confirm that by tracing back how these scores are estimated). In both cases, either we use probabilities or logits, the order of the boxes inside the NMS will not change because as we discussed earlier the two scores are only used for ordering and both should produce the same order (at least that's what the maths say, in practice you might get slight differences due to rounding errors etc). 

----

> What do boxes with negative objectness mean?

The "objectness" are the logit scores estimated by the model for whether there is any object detected to a specific anchor. Logits can be converted to probabilities as said earlier. The higher the probability of an event, the higher its logit (as `p -> 1` then `logit -> +oo` and as `p -> 0` then `logit -> -oo`). Logit < 0.0 means probability < 0.5.

> What makes one box positive while others negative?

The model believes it's more likely an object to be in the anchor. If the probability is > 0.5 it's positive, else negative. Note that this is just a transformation. We don't explicitly remove scores that are positive or negative. In other words the sign of the score does not matter currently.

> Can a totally reasonable box have a negative objectness? Some people in the TVM community are claiming that negative boxes don't matter in the final detection results in practice.

I believe that yes, it's totally possible a reasonable box to have negative objectness. The scale of the score is not taken into account in our implementation and thus it just means that the specific box happened to have a slightly lower score. There is nothing significant about it. If you believe that negative boxes don't matter, then you might do so because you are willing to put a threshold to any box that has probability less than 0.5. But that's something you decide based on the domain, problem, data, cost matrix etc, it's not a general rule.


> If being negative doesn't matter much in terms of objectness, would it be better to rescale them in the range [0, 1), for more intuitive understanding?

Arguably probabilities are easier to interpret than logits but they do come with an overhead because they require additional calculations. Perhaps that extra overhead is justified if people intend to apply thresholds to those probabilities to reduce the number of boxes before the `O(n*logn)` sorting part of NMS. But that's something that needs to be weighted to ensure that it's relatively straightforward to tune and does improve the overall speed without hurting the accuracy. 

To check the above, I drafted PR #3205 to investigate the tradeoff between speed/accuracy if we switch to probabilities and apply thresholds. Note that this is just an investigation, it's definitely not guaranteed that the change will land on master cause it's unclear if there are any benefits.

**Comment by masahi** [comment]
Wow that's a great answer! Thank you very much, it makes a lot of sense now :)

Indeed, since I'm only interested in testing / model compiling, I haven't paid attention to loss functions. But it is a great insight that I can look at the losses to understand what objectness network is trying to calculate.



> Logit < 0.0 means probability < 0.5

:+1: Great now I understand where the negatives come from and that they do matter. 

I'm now definitely convinced that TVM's conversion rule for PyTorch NMS is definitely wrong and needs fixing. I'll try convince others who don't want to remove the "cheat" we are doing because it makes things slower :)

A bit off topic, but I got the impression that you deeply care about the performance of NMS. If that's the case, I suggest reworking your `batched_nms` implementation https://github.com/pytorch/vision/blob/master/torchvision/ops/boxes.py#L80-L88. I think the way it is implemented is more of a hack to get things going fast, but this implementation means most of O(N**2) IOU tests you do in `nms_kernel` and are pointless and also 
O(N**2) memory requirement for the bit mask can become significant for no good reason.

**Comment by datumbox** [comment]
@masahi Awesome! I'll close the ticket but please feel free to open it again if there are more concerns.

Concerning the NMS indeed it's one of the most costly parts of inference so we try to optimize it as much as possible. My understanding is that the current implementation tries to make use of PyTorch specific features to make it very fast, nevertheless if you think there is a better option please send a PR and we will help you benchmark it.

**Comment by fmassa** [comment]
Hi,

Thanks for the very comprehensive discussion @datumbox, I totally agree with your answers.

@masahi about your comment
> reworking your batched_nms implementation

yes, you are right. The current implementation is tailored mostly for GPUs, in order to be able to re-use the current kernels we have and avoid for loops in Python (which would be costly).
It would be preferable to have an optimized C++ / CUDA implementation that could do this more efficiently. For CPU that would most probably mean just a simple for loop, but for CUDA things get a bit more complicated.

**Comment by masahi** [comment]
@fmassa @datumbox 

About `batched_nms` performance, one thing I found is that all usage of `batched_nms` are followed by post NMS topk, like this:

https://github.com/pytorch/vision/blob/3d60f498e71ba63b428edb184c9ac38fa3737fa6/torchvision/models/detection/rpn.py#L261-L263

So if we pass this topk parameter to NMS, we can make NMS faster by returning from the sequential loop below as soon as `num_to_keep` reaches this topk value. 

https://github.com/pytorch/vision/blob/3711754a508e429d0049df3c4a410c4cde08e4e6/torchvision/csrc/ops/cuda/nms_kernel.cu#L148-L160

**Comment by fmassa** [comment]
Yes, that is true, and it's the approach that some kernels follow (like the one from Caffe2 and ONNX).
I'm not necessarily super excited about this change though, as this gives precedence for other micro-optimizations that could be applied to NMS (like the one discussed in https://github.com/pytorch/vision/issues/2552)

But I would like to see some performance numbers on representative use-cases to motivate the potential gains by restricting `num_to_keep` inside NMS -- I wouldn't expect this to make a very significant different runtime-wise on an end-to-end model inference (but I might be wrong)
